src/argilla/server/contexts/datasets.py

```diff
@@ -509,83 +519,190 @@ async def _validate_suggestion(
+        # TODO: We need to remove these asynchronicity requirements somehow
         for record in records:
             await record.awaitable_attrs.responses
+        for vector in all_vectors:
+            await vector.awaitable_attrs.vector_settings
+        # TODO: Review and unify in a single method
         await search_engine.index_records(dataset, records)
+        if all_vectors:
+            await search_engine.set_records_vectors(dataset, all_vectors)
 
     await db.commit()
 
 
+async def validate_record_metadata(
+    db: "AsyncSession",
+    dataset: Dataset,
+    record_idx: int,
+    metadata: Dict[str, Any],
+    cache: Dict[str, MetadataProperty] = {},
+):
+    """Validate metadata for a record."""
+    if not metadata:
+        return
+
+    try:
+        await _validate_metadata(db, dataset=dataset, metadata=metadata, metadata_properties=cache)
+    except ValueError as e:
+        raise ValueError(f"Provided metadata for record at position {record_idx} is not valid: {e}") from e
+
+
+async def create_record_vectors(
+    db: "AsyncSession",
+    dataset: Dataset,
+    record_idx: int,
+    record_create: "RecordCreate",
+    cache: Dict[str, VectorSettingsSchema] = {},
+) -> List[Vector]:
+    """Create vectors for a record."""
+    if not record_create.vectors:
+        return []
+
+    vectors = []
+    for vector_name, vector_value in record_create.vectors.items():
+        try:
+            await validate_vector(db, dataset.id, vector_name, vector_value, vectors_settings=cache)
+            vectors.append(Vector(value=vector_value, vector_settings_id=cache[vector_name].id))
+        except ValueError as e:
+            raise ValueError(
+                f"Provided vector with name={vector_name} of record at position {record_idx} is not valid: {e}"
+            ) from e
+
+    return vectors
+
+
+async def create_record_responses(
+    db: "AsyncSession", dataset: Dataset, record_idx: int, record_create: RecordCreate, cache: Set[UUID] = set()
+) -> List[Response]:
+    """Create responses for a record."""
+    if not record_create.responses:
+        return []
+
+    responses = []
+
+    for idx, response in enumerate(record_create.responses):
+        try:
+            await validate_user(db, response.user_id, cache)
+            validate_response_values(dataset, values=response.values, status=response.status)
+            responses.append(
+                Response(
+                    values=jsonable_encoder(response.values),
+                    status=response.status,
+                    user_id=response.user_id,
+                )
+            )
+        except ValueError as e:
+            raise ValueError(
+                f"Provided response at position {idx} of record at position {record_idx} is not valid: {e}"
+            ) from e
+
+    return responses
+
+
+async def create_record_suggestions(
+    db: "AsyncSession", record_idx: int, record_create: "RecordCreate", cache: Dict[UUID, Question] = {}
```

Comment:
Maybe `record_position` is better? I was thinking that it was a typo.

```suggestion
    db: "AsyncSession", record_position: int, record_create: "RecordCreate", cache: Dict[UUID, Question] = {}
```

Reply:
done